### ISSUE TITLE
Job objects cannot be assigned memory limits greater than 4 GB

### ISSUE THREAD

**Issue by offhub** [sandboxie-plus/Sandboxie#4096]
### Describe what you noticed and did

1. Save the Bat and PS1 codes in separate files.
2. Run the bat file outside the sandbox.
3. Examine the job object memory limits of opened `Wordpad.exe` processes.

<details>
  <summary>issue4096.bat</summary>
  
  ```bat
  @echo off
  
  cls
  
  set lockFile=%TEMP%\lockfile-issue4096
  
  if exist "%lockFile%" (
  	echo Another instance is running.
  	timeout 3
  	exit /b
  ) else (
  	type NUL >"%lockFile%"
  )
  
  set "scriptDir=%~dp0"
  set "sbiePath=%ProgramFiles%\Sandboxie-Plus"
  set "boxName1=TestBox4GB"
  set "boxName2=TestBox4GBPlusOne"
  set "boxName3=TestBox4GBPlusPS"
  set "psFile=jobTest.ps1"
  
  pushd "%cd%"
  cd /d "%sbiePath%"
  
  rem 4 GB
  SbieIni.exe set "%boxName1%" "Enabled" "y"
  SbieIni.exe set "%boxName1%" "TotalMemoryLimit" "4294967295"
  SbieIni.exe set "%boxName1%" "ProcessMemoryLimit" "4294967295"
  SbieIni.exe set "%boxName1%" "ProcessNumberLimit" "20"
  Start.exe /reload
  Start.exe /box:%boxName1% wordpad.exe
  
  rem 4 GB + 1 byte
  SbieIni.exe set "%boxName2%" "Enabled" "y"
  SbieIni.exe set "%boxName2%" "TotalMemoryLimit" "4294967296"
  SbieIni.exe set "%boxName2%" "ProcessMemoryLimit" "4294967296"
  SbieIni.exe set "%boxName2%" "ProcessNumberLimit" "20"
  Start.exe /reload
  Start.exe /box:%boxName2% wordpad.exe
  
  rem 4 GB+ powershell
  if exist "%scriptDir%\%psFile%" (
  	SbieIni.exe set "%boxName3%" "Enabled" "y"
  	SbieIni.exe set "%boxName3%" "NoAddProcessToJob" "y"
  	Start.exe /reload
  	Start.exe /box:%boxName3% powershell.exe -noexit -ExecutionPolicy Bypass -File "%scriptDir%\%psFile%"
  ) else (
  	echo jobTest.ps1 not found, skipping...
  )
  
  echo.
  echo Press any key to exit and remove the boxes...
  echo.
  pause > nul
  
  Start.exe /box:%boxName1% /terminate
  Start.exe /box:%boxName1% delete_sandbox
  SbieIni.exe set %boxName1% * ""
  Start.exe /box:%boxName2% /terminate
  Start.exe /box:%boxName2% delete_sandbox
  SbieIni.exe set %boxName2% * ""
  Start.exe /box:%boxName3% /terminate
  Start.exe /box:%boxName3% delete_sandbox
  SbieIni.exe set %boxName3% * ""
  Start.exe /reload
  
  del "%lockFile%"
  popd
  
  exit /b
  ```

</details>

<details>
  <summary>jobTest.ps1</summary>
  
  ```ps1
  # Load necessary .NET assemblies if not already loaded
  if (-not ([System.Management.Automation.PSTypeName]'JobObjectHelper').Type) {
      Add-Type -TypeDefinition @"
  using System;
  using System.Runtime.InteropServices;
  
  public static class JobObjectHelper
  {
      [DllImport("kernel32.dll", CharSet = CharSet.Unicode, SetLastError = true)]
      public static extern IntPtr CreateJobObject(IntPtr lpJobAttributes, string lpName);
  
      [DllImport("kernel32.dll", SetLastError = true)]
      public static extern bool SetInformationJobObject(IntPtr hJob, int JobObjectInfoClass, IntPtr lpJobObjectInfo, uint cbJobObjectInfoLength);
  
      [DllImport("kernel32.dll", SetLastError = true)]
      public static extern bool AssignProcessToJobObject(IntPtr hJob, IntPtr hProcess);
  
      [DllImport("kernel32.dll", SetLastError = true)]
      public static extern bool CloseHandle(IntPtr hObject);
  
      public const int JobObjectExtendedLimitInformation = 9;
      public const int JobMemoryLimit = 0x00000200;
      public const int ProcessMemoryLimit = 0x00000100;
      public const int ActiveProcessLimitFlag = 0x00000008;
  
      [StructLayout(LayoutKind.Sequential)]
      public struct JOBOBJECT_EXTENDED_LIMIT_INFORMATION
      {
          public JOBOBJECT_BASIC_LIMIT_INFORMATION BasicLimitInformation;
          public IO_COUNTERS IoInfo;
          public IntPtr ProcessMemoryLimit;
          public IntPtr JobMemoryLimit;
          public IntPtr PeakProcessMemoryUsed;
          public IntPtr PeakJobMemoryUsed;
      }
  
      [StructLayout(LayoutKind.Sequential)]
      public struct JOBOBJECT_BASIC_LIMIT_INFORMATION
      {
          public Int64 PerProcessUserTimeLimit;
          public Int64 PerJobUserTimeLimit;
          public uint LimitFlags;
          public IntPtr MinimumWorkingSetSize;
          public IntPtr MaximumWorkingSetSize;
          public uint ActiveProcessLimit;
          public IntPtr Affinity;
          public uint PriorityClass;
          public uint SchedulingClass;
      }
  
      [StructLayout(LayoutKind.Sequential)]
      public struct IO_COUNTERS
      {
          public UInt64 ReadOperationCount;
          public UInt64 WriteOperationCount;
          public UInt64 OtherOperationCount;
          public UInt64 ReadTransferCount;
          public UInt64 WriteTransferCount;
          public UInt64 OtherTransferCount;
      }
  
      public static bool SetJobLimits(IntPtr hJob, long jobMemoryLimitBytes, long processMemoryLimitBytes, uint activeProcessLimit)
      {
          JOBOBJECT_EXTENDED_LIMIT_INFORMATION jobInfo = new JOBOBJECT_EXTENDED_LIMIT_INFORMATION();
          jobInfo.BasicLimitInformation.LimitFlags = JobMemoryLimit | ProcessMemoryLimit | ActiveProcessLimitFlag;
          jobInfo.JobMemoryLimit = new IntPtr(jobMemoryLimitBytes);
          jobInfo.ProcessMemoryLimit = new IntPtr(processMemoryLimitBytes);
          jobInfo.BasicLimitInformation.ActiveProcessLimit = activeProcessLimit;
  
          int jobInfoLength = Marshal.SizeOf(jobInfo);
          IntPtr jobInfoPtr = Marshal.AllocHGlobal(jobInfoLength);
  
          try
          {
              Marshal.StructureToPtr(jobInfo, jobInfoPtr, false);
              return SetInformationJobObject(hJob, JobObjectExtendedLimitInformation, jobInfoPtr, (uint)jobInfoLength);
          }
          finally
          {
              Marshal.FreeHGlobal(jobInfoPtr);
          }
      }
  }
  "@
  }
  
  # Function to set limits for a job object
  function Set-JobLimits {
      param (
          [Parameter(Mandatory=$true)]
          [int]$processId,
          [Parameter(Mandatory=$true)]
          [long]$jobMemoryLimitBytes,
          [Parameter(Mandatory=$true)]
          [long]$processMemoryLimitBytes,
          [Parameter(Mandatory=$true)]
          [uint32]$activeProcessLimit
      )
  
      # Create a job object
      $hJob = [JobObjectHelper]::CreateJobObject([System.IntPtr]::Zero, "PowerShellJob")
  
      if ($hJob -eq [System.IntPtr]::Zero) {
          Write-Host "Failed to create job object. Error: $($([System.Runtime.InteropServices.Marshal]::GetLastWin32Error()))"
          return
      }
  
      # Set job limits
      $success = [JobObjectHelper]::SetJobLimits($hJob, $jobMemoryLimitBytes, $processMemoryLimitBytes, $activeProcessLimit)
      if (-not $success) {
          Write-Host "Failed to set job object information. Error: $($([System.Runtime.InteropServices.Marshal]::GetLastWin32Error()))"
      }
  
      # Assign process to job object
      $processHandle = (Get-Process -Id $processId).Handle
      $success = [JobObjectHelper]::AssignProcessToJobObject($hJob, $processHandle)
      if (-not $success) {
          Write-Host "Failed to assign process to job object. Error: $($([System.Runtime.InteropServices.Marshal]::GetLastWin32Error()))"
      }
  
      # Close the job object handle
      [JobObjectHelper]::CloseHandle($hJob)
  }
  
  # Example usage: Create a new process and set limits for it
  $process = Start-Process -FilePath "wordpad.exe" -PassThru
  Set-JobLimits -processId $process.Id -jobMemoryLimitBytes 14294967296 -processMemoryLimitBytes 14294967296 -activeProcessLimit 20

  ```
  
</details>

![issue4096jbo](https://github.com/user-attachments/assets/f5934d84-8082-4719-996f-ce69299a2672)

### How often did you encounter it so far?

Every time

### Expected behavior

Values above 4 GB should be usable.

### Affected program

Not relevant

### Download link

Not relevant

### Where is the program located?

Not relevant to my request.

### Did the program or any related process close unexpectedly?

No, not at all.

### What version of Sandboxie are you running now?

Sandboxie-Plus 1.14.4 64-bit

### Is it a new installation of Sandboxie?

I recently did a new clean installation.

### In which sandbox type you have this problem?

In a standard isolation sandbox (yellow sandbox icon).

### Can you reproduce this problem on a new empty sandbox?

I can confirm it also on a new empty sandbox.

### What is your Windows edition and version?

Windows 10 Pro 22H2 64-bit (19045.4529) and Windows 11 Pro 23H2 64-bit (22631.3447)

### In which Windows account you have this problem?

A local account (Administrator)., A Microsoft account (Administrator)., An account with UAC protection set to Always notify.

### Please mention any installed security software

Microsoft Windows Defender

### Sandboxie.ini configuration

```ini

[TestBox4GB]
Enabled=y
BlockNetworkFiles=y
RecoverFolder=%{374DE290-123F-4565-9164-39C4925E467B}%
RecoverFolder=%Personal%
RecoverFolder=%Desktop%
BorderColor=#00FFFF,ttl
Template=OpenBluetooth
Template=SkipHook
Template=FileCopy
Template=qWave
Template=BlockPorts
Template=LingerPrograms
Template=AutoRecoverIgnore
ConfigLevel=10
TotalMemoryLimit=4294967295
ProcessMemoryLimit=4294967295
ProcessNumberLimit=20

[TestBox4GBPlusOne]
Enabled=y
BlockNetworkFiles=y
RecoverFolder=%{374DE290-123F-4565-9164-39C4925E467B}%
RecoverFolder=%Personal%
RecoverFolder=%Desktop%
BorderColor=#00FFFF,ttl
Template=OpenBluetooth
Template=SkipHook
Template=FileCopy
Template=qWave
Template=BlockPorts
Template=LingerPrograms
Template=AutoRecoverIgnore
ConfigLevel=10
TotalMemoryLimit=4294967296
ProcessMemoryLimit=4294967296
ProcessNumberLimit=20
```

**Comment by love-code-yeyixiao** [comment]
&nbsp;This is clearly not a problem with the feature itself.Sandboxie uses ULONG as the return value for the numeric type configuration.Converting 4GB to KB results in a value exactly equal to the limit of the ULONG datatype.Exceeding this value results in an overflow.










---Original---
From: ***@***.***&gt;
Date: Mon, Jul 22, 2024 09:16 AM
To: ***@***.***&gt;;
Cc: ***@***.***&gt;;
Subject: [sandboxie-plus/Sandboxie] Job objects cannot be assigned memorylimits greater than 4 GB (Issue #4096)




 
Describe what you noticed and did
  
Save the Bat and PS1 codes in separate files.
 
Run the bat file outside the sandbox.
 
Examine the job object memory limits of opened Wordpad.exe processes.
     issueXXX.bat @echo off cls set lockFile=%TEMP%\lockfile-issueXXXX if exist "%lockFile%" ( 	echo Another instance is running. 	timeout 3 	exit /b ) else ( 	type NUL &gt;"%lockFile%" ) set "scriptDir=%~dp0" set "sbiePath=%ProgramFiles%\Sandboxie-Plus" set "boxName1=TestBox4GB" set "boxName2=TestBox4GBPlusOne" set "boxName3=TestBox4GBPlusPS" set "psFile=jobTest.ps1" pushd "%cd%" cd /d "%sbiePath%" rem 4 GB SbieIni.exe set "%boxName1%" "Enabled" "y" SbieIni.exe set "%boxName1%" "TotalMemoryLimit" "4294967295" SbieIni.exe set "%boxName1%" "ProcessMemoryLimit" "4294967295" SbieIni.exe set "%boxName1%" "ProcessNumberLimit" "20" Start.exe /reload Start.exe /box:%boxName1% wordpad.exe rem 4 GB + 1 byte SbieIni.exe set "%boxName2%" "Enabled" "y" SbieIni.exe set "%boxName2%" "TotalMemoryLimit" "4294967296" SbieIni.exe set "%boxName2%" "ProcessMemoryLimit" "4294967296" SbieIni.exe set "%boxName2%" "ProcessNumberLimit" "20" Start.exe /reload Start.exe /box:%boxName2% wordpad.exe rem 4 GB+ powershell if exist "%scriptDir%\%psFile%" ( 	SbieIni.exe set "%boxName3%" "Enabled" "y" 	SbieIni.exe set "%boxName3%" "NoAddProcessToJob" "y" 	Start.exe /reload 	Start.exe /box:%boxName3% powershell.exe -noexit -ExecutionPolicy Bypass -File "%scriptDir%\%psFile%" ) else ( 	echo jobTest.ps1 not found, skipping... ) echo. echo Press any key to exit and remove the boxes... echo. pause &gt; nul Start.exe /box:%boxName1% /terminate Start.exe /box:%boxName1% delete_sandbox SbieIni.exe set %boxName1% * "" Start.exe /box:%boxName2% /terminate Start.exe /box:%boxName2% delete_sandbox SbieIni.exe set %boxName2% * "" Start.exe /box:%boxName3% /terminate Start.exe /box:%boxName3% delete_sandbox SbieIni.exe set %boxName3% * "" Start.exe /reload del "%lockFile%" popd exit /b
     jobTest.ps1 # Load necessary .NET assemblies if not already loaded if (-not ([System.Management.Automation.PSTypeName]'JobObjectHelper').Type) {     Add-Type -TypeDefinition @" using System; using System.Runtime.InteropServices;  public static class JobObjectHelper {     [DllImport("kernel32.dll", CharSet = CharSet.Unicode, SetLastError = true)]     public static extern IntPtr CreateJobObject(IntPtr lpJobAttributes, string lpName);      [DllImport("kernel32.dll", SetLastError = true)]     public static extern bool SetInformationJobObject(IntPtr hJob, int JobObjectInfoClass, IntPtr lpJobObjectInfo, uint cbJobObjectInfoLength);      [DllImport("kernel32.dll", SetLastError = true)]     public static extern bool AssignProcessToJobObject(IntPtr hJob, IntPtr hProcess);      [DllImport("kernel32.dll", SetLastError = true)]     public static extern bool CloseHandle(IntPtr hObject);      public const int JobObjectExtendedLimitInformation = 9;     public const int JobMemoryLimit = 0x00000200;     public const int ProcessMemoryLimit = 0x00000100;     public const int ActiveProcessLimitFlag = 0x00000008;      [StructLayout(LayoutKind.Sequential)]     public struct JOBOBJECT_EXTENDED_LIMIT_INFORMATION     {         public JOBOBJECT_BASIC_LIMIT_INFORMATION BasicLimitInformation;         public IO_COUNTERS IoInfo;         public IntPtr ProcessMemoryLimit;         public IntPtr JobMemoryLimit;         public IntPtr PeakProcessMemoryUsed;         public IntPtr PeakJobMemoryUsed;     }      [StructLayout(LayoutKind.Sequential)]     public struct JOBOBJECT_BASIC_LIMIT_INFORMATION     {         public Int64 PerProcessUserTimeLimit;         public Int64 PerJobUserTimeLimit;         public uint LimitFlags;         public IntPtr MinimumWorkingSetSize;         public IntPtr MaximumWorkingSetSize;         public uint ActiveProcessLimit;         public IntPtr Affinity;         public uint PriorityClass;         public uint SchedulingClass;     }      [StructLayout(LayoutKind.Sequential)]     public struct IO_COUNTERS     {         public UInt64 ReadOperationCount;         public UInt64 WriteOperationCount;         public UInt64 OtherOperationCount;         public UInt64 ReadTransferCount;         public UInt64 WriteTransferCount;         public UInt64 OtherTransferCount;     }      public static bool SetJobLimits(IntPtr hJob, long jobMemoryLimitBytes, long processMemoryLimitBytes, uint activeProcessLimit)     {         JOBOBJECT_EXTENDED_LIMIT_INFORMATION jobInfo = new JOBOBJECT_EXTENDED_LIMIT_INFORMATION();         jobInfo.BasicLimitInformation.LimitFlags = JobMemoryLimit | ProcessMemoryLimit | ActiveProcessLimitFlag;         jobInfo.JobMemoryLimit = new IntPtr(jobMemoryLimitBytes);         jobInfo.ProcessMemoryLimit = new IntPtr(processMemoryLimitBytes);         jobInfo.BasicLimitInformation.ActiveProcessLimit = activeProcessLimit;          int jobInfoLength = Marshal.SizeOf(jobInfo);         IntPtr jobInfoPtr = Marshal.AllocHGlobal(jobInfoLength);          try         {             Marshal.StructureToPtr(jobInfo, jobInfoPtr, false);             return SetInformationJobObject(hJob, JobObjectExtendedLimitInformation, jobInfoPtr, (uint)jobInfoLength);         }         finally         {             Marshal.FreeHGlobal(jobInfoPtr);         }     } } "@ } # Function to set limits for a job object function Set-JobLimits {     param (         [Parameter(Mandatory=$true)]         [int]$processId,         [Parameter(Mandatory=$true)]         [long]$jobMemoryLimitBytes,         [Parameter(Mandatory=$true)]         [long]$processMemoryLimitBytes,         [Parameter(Mandatory=$true)]         [uint32]$activeProcessLimit     )     # Create a job object     $hJob = [JobObjectHelper]::CreateJobObject([System.IntPtr]::Zero, "PowerShellJob")     if ($hJob -eq [System.IntPtr]::Zero) {         Write-Host "Failed to create job object. Error: $($([System.Runtime.InteropServices.Marshal]::GetLastWin32Error()))"         return     }     # Set job limits     $success = [JobObjectHelper]::SetJobLimits($hJob, $jobMemoryLimitBytes, $processMemoryLimitBytes, $activeProcessLimit)     if (-not $success) {         Write-Host "Failed to set job object information. Error: $($([System.Runtime.InteropServices.Marshal]::GetLastWin32Error()))"     }     # Assign process to job object     $processHandle = (Get-Process -Id $processId).Handle     $success = [JobObjectHelper]::AssignProcessToJobObject($hJob, $processHandle)     if (-not $success) {         Write-Host "Failed to assign process to job object. Error: $($([System.Runtime.InteropServices.Marshal]::GetLastWin32Error()))"     }     # Close the job object handle     [JobObjectHelper]::CloseHandle($hJob) } # Example usage: Create a new process and set limits for it $process = Start-Process -FilePath "wordpad.exe" -PassThru Set-JobLimits -processId $process.Id -jobMemoryLimitBytes 14294967296 -processMemoryLimitBytes 14294967296 -activeProcessLimit 20 
  
How often did you encounter it so far?
 
Every time
 
Expected behavior
 
Values above 4 GB should be usable.
 
Affected program
 
Not relevant
 
Download link
 
Not relevant
 
Where is the program located?
 
Not relevant to my request.
 
Did the program or any related process close unexpectedly?
 
No, not at all.
 
Crash dump
 
No response
 
What version of Sandboxie are you running now?
 
Sandboxie-Plus 1.14.4 64-bit
 
Is it a new installation of Sandboxie?
 
I recently did a new clean installation.
 
Is it a regression from previous versions?
 
No response
 
In which sandbox type you have this problem?
 
In a standard isolation sandbox (yellow sandbox icon).
 
Can you reproduce this problem on a new empty sandbox?
 
I can confirm it also on a new empty sandbox.
 
What is your Windows edition and version?
 
Windows 10 Pro 22H2 64-bit (19045.4529) and Windows 11 Pro 23H2 64-bit (22631.3447)
 
In which Windows account you have this problem?
 
A local account (Administrator)., A Microsoft account (Administrator)., An account with UAC protection set to Always notify.
 
Please mention any installed security software
 
Microsoft Windows Defender
 
Did you previously enable some security policy settings outside Sandboxie?
 
No response
 
Trace log
 
No response
 
Sandboxie.ini configuration
 Enabled=y BlockNetworkFiles=y RecoverFolder=%{374DE290-123F-4565-9164-39C4925E467B}% RecoverFolder=%Personal% RecoverFolder=%Desktop% BorderColor=#00FFFF,ttl Template=OpenBluetooth Template=SkipHook Template=FileCopy Template=qWave Template=BlockPorts Template=LingerPrograms Template=AutoRecoverIgnore ConfigLevel=10 TotalMemoryLimit=4294967295 ProcessMemoryLimit=4294967295 ProcessNumberLimit=20 Enabled=y BlockNetworkFiles=y RecoverFolder=%{374DE290-123F-4565-9164-39C4925E467B}% RecoverFolder=%Personal% RecoverFolder=%Desktop% BorderColor=#00FFFF,ttl Template=OpenBluetooth Template=SkipHook Template=FileCopy Template=qWave Template=BlockPorts Template=LingerPrograms Template=AutoRecoverIgnore ConfigLevel=10 TotalMemoryLimit=4294967296 ProcessMemoryLimit=4294967296 ProcessNumberLimit=20

**Comment by DavidXanatos** [comment]
Yes, the bug was with the value conversion, I have added a SbieApi_QueryConfNumber64 which returns a ULONG64, used it instead and it should now work fine

**Comment by DavidXanatos** [comment]
LOL a 4 GiB bag in a 4096 our 4 KiB issue what a coincidence

**Comment by wilders-soccerfan** [comment]
The Box UI in v1.14.4 shows limits for single and total memory as "Unit:KB". The actual units are bytes. 
Perhaps this UI should also be fixed in the upcoming v1.4.5.
![4096](https://github.com/user-attachments/assets/da2619f5-f4a9-4234-b52f-ec98baa467a1)
